RFC/rfcBBL209/README.md

```diff
@@ -0,0 +1,104 @@
+
+#### Security
+
+In order for this scheme to be secure it must be true that there is only a single pair `(S_i-1, P_i)` that can be produced to match with `S_i`. If the pair must be of the form `(S_i-1, P_malicious)` then this is certainly true since otherwise one could create a collision on the overall hash function. However, given that there are two parameters to vary it seems possible this could be computationally easier than finding a collision on the overall hash function.
+
+#### SHA-3
+
+While SHA-3 is not a Merkle–Damgård construction it follows the same psuedocode structure above
+
+### Example: Tree constructions like Blake3, Kangaroo-Twelve, or ParallelHash
+
+In tree constructions we are not restricted to downloading the file backwards and can instead download the parts of the file the we are looking for, which includes downloading the file forwards for sequential streaming.
+
+There is detail about how to do this for Blake3 in the [Blake3 paper](https://github.com/BLAKE3-team/BLAKE3-specs/blob/master/blake3.pdf) section 6.4, Verified Streaming
+
+### Implementation Plan
+
+#### Bitswap changes
+
+* When a server responds to a request for a block if the block is too large then instead send a traversal order list of the block as defined by the particular hash function used (e.g. linear and backwards for SHA-1,2,3)
+* Large Manifests
+  * If the list is more than 1MiB long then only send the first 1MiB along with an indicator that the manifest is not complete
+  * When the client is ready to process more of the manifest then it can send a request WANT_LARGE_BLOCK_MANIFEST containing the multihash of the entire large block and the last hash in the manifest
+* When requesting subblocks send requests as `(full block multihash, start index, end index)`
+  * process subblock responses separately from full block responses verifying the results as they come in
+* As in [RFC|BB|L2 - Speed up Bitswap with GraphSync CID only query](https://github.com/protocol/beyond-bitswap/issues/25) specify how much trust goes into a given manifest, examples include
+  * download at most 20 unverified blocks at a time from a given manifest
+  * grow trust geometrically (e.g. 10 blocks, then if those are good 20, 40, ...)
+
+#### Datastore
+
+* Servers should cache/store a particular chunking for the traversal that is defined by the implementation for the particular hash function (e.g. 256 KiB segments for SHA-2)
+  * Once clients receive the full block they should process it and store the chunking, reusing the work from validating the block
+* Clients and servers should have a way of aliasing large blocks as a concatenated set of smaller blocks
+* Need to quarantine subblocks until the full block is verified as in [RFC|BB|L2 - Speed up Bitswap with GraphSync CID only query](https://github.com/protocol/beyond-bitswap/issues/25)
+
+#### Hash function support
+
+* Add support for SHA-1/2 (should be very close to the same)
```

Comment:
sha 1 is deprecated / not-recommended at this point. It seems unclear it's valuable or safe to support it. why do we want to?

Reply:
"git"

Reply:
Mostly for Git support, however with Git eventually moving to SHA-2 if it turned out SHA-1 was unworkable we could probably deal with it.

Reply:
I would strongly prefer if support for this for sha-1 is opt-in, if not on by default. See https://docs.softwareheritage.org/devel/swh-model/persistent-identifiers.html which I propose adding in https://github.com/multiformats/multicodec/pull/203 for why.